dbms/src/Flash/ResourceControl/LocalAdmissionController.cpp

```diff
@@ -13,19 +13,491 @@
+    last_fetch_tokens_from_gac_timepoint = std::chrono::steady_clock::now();
+
+    if (resp.responses().empty())
+    {
+        handleBackgroundError("got empty TokenBuckets resp from GAC");
+        return {};
+    }
+
+    const auto now = std::chrono::steady_clock::now();
+    for (const resource_manager::TokenBucketResponse & one_resp : resp.responses())
+    {
+        // For each resource group.
+        if unlikely (!one_resp.granted_resource_tokens().empty())
+        {
+            handleBackgroundError("GAC return RAW granted tokens, but LAC expect RU tokens");
+            continue;
+        }
+
+        handled_resource_group_names.emplace_back(one_resp.resource_group_name());
+        for (const resource_manager::GrantedRUTokenBucket & granted_token_bucket : one_resp.granted_r_u_tokens())
+        {
+            // For each granted token bucket.
+            if unlikely (granted_token_bucket.type() != resource_manager::RequestUnitType::RU)
+            {
+                handleBackgroundError("unexpected request type");
+                continue;
+            }
+
+            int64_t trickle_ms = granted_token_bucket.trickle_time_ms();
+            RUNTIME_CHECK(trickle_ms >= 0);
+
+            double added_tokens = granted_token_bucket.granted_tokens().tokens();
+            RUNTIME_CHECK(added_tokens >= 0);
+
+            int64_t capacity = granted_token_bucket.granted_tokens().settings().burst_limit();
+
+            // Check GAC code, fill_rate is never setted.
+            RUNTIME_CHECK(granted_token_bucket.granted_tokens().settings().fill_rate() == 0);
 
-LocalAdmissionController::LocalAdmissionController()
+            auto resource_group = findResourceGroup(one_resp.resource_group_name());
+
+            if (resource_group == nullptr)
+                continue;
+
+            if (trickle_ms == 0)
+            {
+                // GAC has enough tokens for LAC.
+                resource_group->updateNormalMode(added_tokens, capacity);
+            }
+            else
+            {
+                // GAC doesn't have enough tokens for LAC, start to trickle.
+                resource_group->updateTrickleMode(added_tokens, capacity, trickle_ms);
+            }
+            resource_group->updateFetchTokenTimepoint(now);
+        }
+    }
+    return handled_resource_group_names;
+}
+
+void LocalAdmissionController::watchGAC()
 {
-    if (!global_instance)
-        global_instance = std::make_unique<MockLocalAdmissionController>();
+    while (true)
+    {
+        doWatch();
+
+        // Got here when:
+        // 1. grpc stream write error.
+        // 2. grpc stream read error.
+        // 3. watch is cancel.
+        // Will sleep and try again.
+        std::unique_lock lock(mu);
+        if (cv.wait_for(lock, std::chrono::seconds(30), [this]() { return stopped.load(); }))
+            return;
+    }
 }
 
-LocalAdmissionController::~LocalAdmissionController()
+void LocalAdmissionController::doWatch()
```

Comment:
Why is watch gac needed here?

Reply:
Isn’t the fetch resource group and refill bucket token pinging the GAC?

Reply:
`watchGAC` is to watch the `deletion` and `config change` of resource group.

We can fetch these info from GAC periodically, but I think it's no accurate. And tidb/tikv also watch etcd for resource group creation/deletion/change

Reply:
ok